pcicrawler/ocp_diag.py

```diff
@@ -0,0 +1,262 @@
+        return super().run(self._device.express_aer)
+
+
+@dataclass
+class TestSet:
+    '''Object to hold a list of DUTs to be tested and 
+    the values they need to be compared against.'''
+    duts: list
+    conditions: dict
+
+    def __str__(self):
+        return str(self.duts)
+
+
+# List of available tests to iterate through
+AVAILABLE_TESTS = {
+    'location': CheckLocation,
+    'vendor_id': CheckVendorID,
+    'device_id': CheckDeviceID,
+    'class_id': CheckClassID,
+    'physical_slot': CheckPhysicalSlot,
+    'current_link_speed': CheckCurrentLinkSpeed,
+    'current_link_width': CheckCurrentLinkWidth,
+    'capable_link_speed': CheckCapableLinkSpeed,
+    'capable_link_width': CheckCapableLinkWidth,
+    'addresses': CheckAddress,
+    'check_aer': CheckAER,
+}
+
+
+class OCPOutputObj:
+    def __init__(self,
+                 devs: ty.Set[PCIDevice],
+                 json_path: str,
+                 ocp_run: tv.TestRun):
+        self._json_path = json_path
+        self._devs = devs
+        self._ocp_run = ocp_run
+        self._result = True
+
+    def run(self):
+        duts = []
+        # Process input json
+        with open(self._json_path, 'r') as fh:
+            input_json = json.load(fh)
+            try:
+                duts = input_json['duts']
+            except KeyError as _:
+                raise json.JSONDecodeError("ERROR: Cannot parse file {}, missing DUTs info.".format(self._json_path))
+        for dut in duts:
+            for identifier in dut['identifiers']:
+                if type(dut['identifiers'][identifier]) != str:
+                    dut['identifiers'][identifier] = str(dut['identifiers'][identifier])
+        test_sets = []
+        for dut in duts:
+            identifier = dut['identifiers']
+            potential_duts = list(self._devs)
+            # Go look for this device and get the data
+            missing_devices = []
+            if 'address' in identifier:
+                if identifier['address'] not in potential_duts:
+                    missing_devices.append("No device was found with address {}.".format(identifier['address']))
+                potential_duts = [dut for dut in potential_duts if str(dut) == identifier['address']]
+            if 'vendor_id' in identifier:
+                potential_duts = [dut for dut in potential_duts if dut.vendor_id == identifier['vendor_id']]
+                if len(potential_duts) == 0:
+                    missing_devices.append("No device was found with vendor id {}.".format(identifier['vendor_id']))
+            if 'device_id' in identifier:
+                potential_duts = [dut for dut in potential_duts if dut.device_id == identifier['device_id']]
+                if len(potential_duts) == 0:
+                    missing_devices.append("No device was found with device id {}.".format(identifier['device_id']))
+            if len(potential_duts) != 0:
+                test_sets.append(TestSet(duts=potential_duts,
```

Comment:
suggestion for the enum type, instead of
`identifier={"address": "some addr"}`
you should have (some serialization of an enum)
`identifier_type="address", identifier_value="some addr"`

this then refactors all of your ifs to a single switch on `identifier_type` which cannot have collisions

Reply:
alternatively you can keep the current serialization (maybe rename `identifiers` to `identifier`) and error if multiple types of identifiers are present in that dict (which makes it back into an enum)